test/controllers/samples_controller_test.rb

```diff
@@ -178,4 +184,153 @@ class SamplesControllerTest < ActionDispatch::IntegrationTest
+  end
+
+  test 'get metadata_fields non-admin user own sample' do
+    post user_session_path, params: @user_nonadmin_params
+
+    get metadata_fields_samples_url, params: {
+      sampleIds: [@sample_human_existing_metadata_joe_project.id]
+    }
+    assert_response :success
+
+    assert_equal ["Nucleotide Type", "Age", "Sex", "Sample Type", "Admission Date"], @response.parsed_body.pluck("name")
+  end
+
+  test 'get metadata_fields non-admin user private sample' do
+    post user_session_path, params: @user_nonadmin_params
+
+    assert_raises(ActiveRecord::RecordNotFound) do
+      get metadata_fields_samples_url, params: {
+        sampleIds: [@metadata_validation_sample_human_existing_metadata.id]
+      }
+    end
+  end
+
+  # If multiple samples, merge the fields.
+  test 'get metadata_fields non-admin user multiple samples' do
+    post user_session_path, params: @user_nonadmin_params
+
+    get metadata_fields_samples_url, params: {
+      sampleIds: [@sample_human_existing_metadata_public.id, @sample_human_existing_metadata_expired.id]
+    }
+    assert_response :success
+
+    assert_equal ["Age", "Nucleotide Type", "Sample Type"], @response.parsed_body.pluck("name")
+  end
+
+  # If multiple samples but one is invalid, return fields for the valid ones.
+  test 'get metadata_fields non-admin user multiple samples one private' do
+    post user_session_path, params: @user_nonadmin_params
+
+    get metadata_fields_samples_url, params: {
+      sampleIds: [@sample_human_existing_metadata_public.id, @metadata_validation_sample_human_existing_metadata.id]
+    }
+    assert_response :success
+
+    assert_equal ["Nucleotide Type", "Sample Type"], @response.parsed_body.pluck("name")
+  end
+
+  test 'save_metadata_fields non-admin user public sample' do
+    post user_session_path, params: @user_nonadmin_params
+
+    assert_raises(ActiveRecord::RecordNotFound) do
+      post save_metadata_v2_sample_url(@sample_human_existing_metadata_public), params: {
+        field: "sample_type",
+        value: "Foobar Sample Type"
+      }
+    end
+  end
+
+  test 'save_metadata_fields non-admin user expired sample' do
+    post user_session_path, params: @user_nonadmin_params
+
+    assert_raises(ActiveRecord::RecordNotFound) do
+      post save_metadata_v2_sample_url(@sample_human_existing_metadata_expired), params: {
+        field: "sample_type",
+        value: "Foobar Sample Type"
+      }
+    end
+  end
+
+  test 'save_metadata_fields non-admin user own sample' do
+    post user_session_path, params: @user_nonadmin_params
+
+    post save_metadata_v2_sample_url(@sample_human_existing_metadata_joe_project), params: {
+      field: "sample_type",
+      value: "Foobar Sample Type"
+    }
+
+    assert_response :success
+
+    assert_equal "Foobar Sample Type", @sample_human_existing_metadata_joe_project.metadata.find_by(key: "sample_type").raw_value
+  end
+
+  test 'save_metadata_fields non-admin user private sample' do
```

Comment:
could we stick with our name convention like "joe cannot save metadata fields to private sample"? :) It makes it a lot easier to see what's going on and what the implications are if a particular test fails

Reply:
Sure.

Reply:
Fixed. It does look much better!